Add more C lib methods

Tokazama opened the pull request.

Tokazama committed:
Add more C lib methods

vtjnash reviewed (changes requested):
Can you convert the existing uses first to use these, so we can use these to reduce duplication?

Tokazama committed:
Replace ccall's with Libc methods

Some ccalls remain in order to accomodate bootstraps still

quinnj commented:
Looks like there's still an import problem in Random? 
```
UndefVarError: `memcpy` not defined
 ```

vchuravy commented on base/libc.jl:
```diff
@@ -371,6 +372,42 @@ Call `calloc` from the C standard library.
 calloc(num::Integer, size::Integer) = ccall(:calloc, Ptr{Cvoid}, (Csize_t, Csize_t), num, size)
 
+"""
+    memcpy(dst::Ptr, src::Ptr, n::Integer) -> Ptr{Cvoid}
```
Need a Julia compat annotation

Tokazama committed:
Add import statement and compat notes

Tokazama commented:
Just a heads up, I'm still getting some test errors locally.

vchuravy commented on base/libc.jl:
```diff
@@ -371,6 +372,58 @@ Call `calloc` from the C standard library.
+
+Call `memcpy` from the C standard library.
+
+!!! compat "Julia 1.9"
```
This will be merged for `1.10` not `1.9`

fingolfin commented on base/libc.jl:
```diff
@@ -371,6 +372,58 @@ Call `calloc` from the C standard library.
+Call `memcpy` from the C standard library.
+
+!!! compat "Julia 1.9"
+    Support for `memcpy` as a tuple requires at least Julia 1.9.
```
What does "as a tuple" mean in this context?

Tokazama committed:
Fix comapt notes

Tokazama committed:
Fix imports and general adoption of C fxns

Tokazama commented:
I think I caught all the import and got testing to work locally at this point.

vtjnash commented on stdlib/Random/src/XoshiroSimd.jl:
```diff
@@ -180,7 +181,7 @@ end
         ref = Ref(f(res, T))
         # TODO: This may make the random-stream dependent on system endianness
-        ccall(:memcpy, Ptr{Cvoid}, (Ptr{UInt8}, Ptr{UInt64}, Csize_t), dst+i, ref, len-i)
+        memcpy(dst+i, pointer_from_objref(ref), len-i)
```
```suggestion
        GC.@preserve ref memcpy(dst+i, pointer_from_objref(ref), len-i)
```

vtjnash commented on stdlib/Random/src/XoshiroSimd.jl:
```diff
@@ -222,7 +223,7 @@ end
         resLoc = _and(res, 0x0101010101010101)
         ref = Ref(resLoc)
-        ccall(:memcpy, Ptr{Cvoid}, (Ptr{UInt8}, Ptr{UInt64}, Csize_t), dst+i, ref, len-i)
+        memcpy(dst+i, pointer_from_objref(ref), len-i)
```
```suggestion
        GC.@preserve ref memcpy(dst+i, pointer_from_objref(ref), len-i)
```

Tokazama committed:
Update stdlib/Random/src/XoshiroSimd.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

Tokazama committed:
Update stdlib/Random/src/XoshiroSimd.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

Tokazama committed:
Add news

Tokazama commented:
Should these methods be usable from the native compiler since "array.jl" and "substring.jl" use them or do we want to keep the native compiler as small as possible here?

Tokazama committed:
Make C-memory calls available from bootstrapping

The Libc module contains some code that we might not care to have as
part of bootstrapping. However, the C-memory methods are directly called
throughout bootstrapping so these are now defined in a seperate
"cmem.jl" file that is defined in Base then imported into `Libc` for the
public interface.

Tokazama committed:
Fix DSFMT memcpy

Tokazama committed:
Fix clipboard.jl memcpy

Tokazama committed:
Replace pointer_to_objref

pointer_to_objref(::Array) is not same as unsafe_convert(Ptr, ::Array)

vtjnash commented on base/array.jl:
```diff
@@ -1835,22 +1835,60 @@ function empty!(a::Vector)
 end
 
 _memcmp(a, b, len) = ccall(:memcmp, Cint, (Ptr{Cvoid}, Ptr{Cvoid}, Csize_t), a, b, len % Csize_t) % Int
```
`bitset.jl` and `parse.jl` seem to be the only remaining users of this, and look trivial to change to `memset`?

vtjnash commented on stdlib/Random/src/Random.jl:
```diff
@@ -16,7 +16,7 @@ using Base.GMP: Limb
 
 using Base: BitInteger, BitInteger_types, BitUnsigned, require_one_based_indexing
-
+using Base.Libc
```
seems unused? might be good to be precise here though:
```suggestion
using Base.Libc: memcpy
```

vtjnash commented on base/libc.jl:
```diff
@@ -335,45 +338,6 @@ if Sys.iswindows()
     end
 end
 
-## Memory related ##
```
These don't appear to be necessary to move. They therefore should stay here.

vtjnash replied:
We don't appear to need any of the malloc/free/etc you removed from here, just the memcpy/memmove/memset/memcmp you added

Tokazama replied:
Ah, I see. I was just trying to keep the 'mem' stuff together, but I can put the other ones back

Tokazama committed:
Update base/array.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

Tokazama committed:
Replace remaing `ccal(:memcmp, ...`

Tokazama committed:
Move c-mem fxns back

vchuravy commented on stdlib/Random/src/XoshiroSimd.jl:
```diff
@@ -222,7 +223,7 @@ end
         resLoc = _and(res, 0x0101010101010101)
         ref = Ref(resLoc)
-        ccall(:memcpy, Ptr{Cvoid}, (Ptr{UInt8}, Ptr{UInt64}, Csize_t), dst+i, ref, len-i)
+        GC.@preserve ref memcpy(dst+i, Base.unsafe_convert(Ptr{Cvoid}, ref), len-i)
```
Why the explicit `@preserve` and `unsafe_convert`?

There is a `cconvert`/`unsafe_convert` chaing defined for `Ptr{Cvoid}`/`Ref{T}`

`Base.unsafe_convert(Ptr{Cvoid}, Base.cconvert(Ptr{Cvoid}, Ref{Int}()))`

Tokazama replied:
the `mem*` functions only accept pointers so there's the explicit conversion to a pointer there. I figured we wanted those methods to be consistent with the other `Libc` memory functions that typically only accept pointers.

PallHaraldsson commented:
Before 1.10 is release shouldn't those ideally be named memcpy!, memset!, memmove!, memcmp!, to conform to Julia naming conversion (even though not used in C)? I don't think we should commit to supporting without !. If that is needed/wanted we should support as synonym, or at least if this is forgotten and 1.10 released, but I would prefer to not have both.

If changed remember to backport to 1.10.

And about the GC_reserve stuff, I'm not sure, but should it be defined in those functions not their callers?

Tokazama commented:
The methods here are barebones bindings to C and don't keep with the norms in most Julia code. These methods don't reflect Julia conventions in name or design and probably can't fully follow Julia's conventions. For example, I'm not sure how we'd appropriately make `memcpy` and `memmove` more "Julian". We could replace `memcpy` with `unsafe_copyto!(dst::Ptr{UInt8}, src::Ptr{UInt8}, n::Integer)` but if there might be overlapping memory it would be better for that to be backed by `memmove`. There's also the question of how we would handle existing methods like `Libc.realloc` and `Libc.free` that also change memory when called. 

My personal understanding is the code here only exists to help move C functionality into Julia code, not as something people should use under most circumstances. I'm not sure at what point we care more about being consistent with Julia's conventions or reflecting the C method so it's easy to find and use.

PallHaraldsson commented:
You have a point with realloc already there. It and these new are mutating, and maybe should be known, at least to all using C. We can keep C's naming, there are pros and cons, but maybe then explicitly document as mutating, and breaking Julia convention? Is it too excessive to do it on all the functions, rather than somewhere globally?

Tokazama commented:
It would probably be a good idea to document that these methods are generally considered unsafe memory operations and also document which ones mutate memory. 

> If this where used from Julia code, then some users would be blind to the mutation, so do you support changing; also realloc -> realloc!, but keep older names as a synonym (to not break; still deprecate?)? Maybe only a synonym for realloc, since 1.10 is not stable yet...

At this point the syntax is consistent at least within the `Libc` module itself. If we're going to change any of the syntax here based on mutation and safety then we should probably discuss renaming most of the functions here. Otherwise, the syntax isn't entirely consistent with Julia code or C code.

This might be worth bringing up in triage because it should probably get more input than me.

Tokazama commented:
There's a triage channel on slack and I think the zoom call is on scheduled+linked on the community calendar
